### ISSUE TITLE
[Datasets] `Dataset.map_batches` progress bar is incorrect

### ISSUE THREAD

**Issue by v4if** [ray-project/ray#33517]
### What happened + What you expected to happen

iterator over the dataset with batch_size, the progress bar is wrong. The total amount of the progress bar is the size of the row, not the size of the batch. its 100, not 10.

logs:
```text
2023-03-21 20:17:23,818	INFO worker.py:1243 -- Using address 127.0.0.1:6379 set in the environment variable RAY_ADDRESS
2023-03-21 20:17:23,819	INFO worker.py:1364 -- Connecting to existing Ray cluster at address: 172.18.3.79:6379...
2023-03-21 20:17:23,824	INFO worker.py:1550 -- Connected to Ray cluster. View the dashboard at http://172.18.3.79:8265
Dataset(num_blocks=100, num_rows=100, schema=<class 'int'>)
2023-03-21 20:17:26,015	INFO bulk_executor.py:39 -- Executing DAG InputDataBuffer[Input] -> TaskPoolMapOperator[read->MapBatches(process_sample)]
read->MapBatches(process_sample):   0%|                                                                                               | 0/100 [00:00<?, ?it/s](_map_task pid=4317, ip=172.16.3.23) process_sample: 10
read->MapBatches(process_sample):   1%|▊                                                                                      | 1/100 [00:00<00:09,  9.94it/s](_map_task pid=4317, ip=172.16.3.23) process_sample: 10
read->MapBatches(process_sample):   2%|█▋                                                                                     | 2/100 [00:01<00:56,  1.75it/s](_map_task pid=4316, ip=172.16.3.23) process_sample: 10
(_map_task pid=4464, ip=172.16.3.23) process_sample: 10
(_map_task pid=4453, ip=172.16.3.23) process_sample: 10
read->MapBatches(process_sample):   3%|██▌                                                                                    | 3/100 [00:01<00:56,  1.71it/s](_map_task pid=4473, ip=172.16.3.23) process_sample: 10
(_map_task pid=4467, ip=172.16.3.23) process_sample: 10
(_map_task pid=4462, ip=172.16.3.23) process_sample: 10
(_map_task pid=4476, ip=172.16.3.23) process_sample: 10
read->MapBatches(process_sample):  10%|████████▌                                                                             | 10/100 [00:01<00:15,  5.87it/s]
[100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199]
```

### Versions / Dependencies

ray, version 2.3.0

### Reproduction script

```python
import ray
from ray.data.read_api import read_datasource
from ray.data.datasource.datasource import Datasource, Reader, ReadTask
from typing import Any, Dict, List, Optional, Union
from ray.data.block import Block, BlockMetadata

ray.init()

class SampleDatasourceReader(Reader):
    def __init__(self, sample_count):
        self.sample_count = sample_count

    def estimate_inmemory_data_size(self) -> Optional[int]:
        """Return an estimate of the in-memory data size, or None if unknown.

        Note that the in-memory data size may be larger than the on-disk data size.
        """
        return None

    def _read_file(self, idx) -> List[Block]:
        blocks = []
        block = [idx]
        blocks.append(block)
        return blocks

    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
        read_tasks: List[ReadTask] = []

        for i in range(self.sample_count):
            metadata = BlockMetadata(
                num_rows=1,
                size_bytes=10, # guess
                schema=None,
                input_files=None,
                exec_stats=None,
            )

            read_task = ReadTask(lambda idx=i: self._read_file(idx), metadata)
            read_tasks.append(read_task)

        return read_tasks


class SampleDatasource(Datasource):
    def create_reader(self, sample_count, **read_args) -> "Reader[T]":
        return SampleDatasourceReader(sample_count)


def read_sample(
    sample_count: int,
    parallelism: int = -1,
    ray_remote_args: Dict[str, Any] = None,
    **read_args,
):
    return read_datasource(
        SampleDatasource(),
        parallelism=parallelism,
        ray_remote_args=ray_remote_args,
        sample_count=sample_count,
        **read_args,
    )

def process_sample(batch):
    print("process_sample:", len(batch))
    return [d + 100 for d in batch]

ds = read_sample(100)
print(ds)
ds = ds.map_batches(
    process_sample,
    batch_size=10
)

result = []
for batch in ds.iter_batches(batch_size=10):
    result = result + batch

print(result)
ray.shutdown()
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by bveeramani** [comment]
```
import time

import ray

dataset = ray.data.range_table(10_000)
dataset = dataset.repartition(1_000)


def identity(batch):
    time.sleep(1)
    return batch


dataset.map_batches(identity, batch_size=100).cache()
```

```
2023-03-22 13:34:48,474 INFO bulk_executor.py:42 -- Executing DAG InputDataBuffer[Input] -> TaskPoolMapOperator[ReadRange] -> AllToAllOperator[Repartition] -> TaskPoolMapOperator[MapBatches(identity)]
ReadRange 0:  10%|█████████████████████▎                                                                                                                                                                                               | 2/20 [00:00<00:06,  2.86it/s]
Repartition 0:   0%|▏                                                                                                                                                                                                                | 1/1000 [00:00<07:39,  2.18it/s]
MapBatches(identity) 0:   9%|██████████████████                                                                                                                                                                                     | 91/1000 [00:10<01:32,  9.85it/s]
```

Seems like there're two issues:
1. Progress bars don't reach 100%.
2. `map_batches` progress bar counts batches instead of blocks (it's showing ~91 batches processed out of 1000 blocks).

cc @ericl I think you were recently working with progress bars?

**Comment by bveeramani** [comment]
Related: https://github.com/ray-project/ray/issues/33317

**Comment by ericl** [comment]
I think this is since setting `batch_size` will dynamically change the number of tasks (reducing the number of tasks). We could probably dynamically adjust the progress bar here in this case as well.

**Comment by v4if** [comment]
> I think this is since setting `batch_size` will dynamically change the number of tasks (reducing the number of tasks). We could probably dynamically adjust the progress bar total here in this case as well.
> 
> cc @c21

Is it planned to release in the version 2.5?

**Comment by raulchen** [comment]
@v4if This should be fixed by #36679 and will be available in 2.6
